### PR TITLE
T5900 dns forwarding: reliability improvements

### PR DESCRIPTION
**Pull request by fvlaicu**
## Change Summary

dns forwarding: [T5900](https://vyos.dev/T5900)
implement in vyos the powerdns dont-throttle-netmasks and server-stale-extensions options.
see https://docs.powerdns.com/recursor/appendices/internals.html#serve-stale about serving stale records.
this will transform the dns cache into a LRU cache

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5900

## Component(s) name

dns forwarder
## Smoketest result

```
root@vyos:~# /usr/libexec/vyos/tests/smoke/cli/test_service_dns_forwarding.py
test_basic_forwarding (__main__.TestServicePowerDNS.test_basic_forwarding) ...
DNS forwarding requires a listen-address


DNS forwarding requires a listen-address

ok
test_dns64 (__main__.TestServicePowerDNS.test_dns64) ...
DNS 6to4 prefix must be of length /96

ok
test_dnssec (__main__.TestServicePowerDNS.test_dnssec) ... ok
test_domain_forwarding (__main__.TestServicePowerDNS.test_domain_forwarding) ... ok
test_dont_throttle_netmasks (__main__.TestServicePowerDNS.test_dont_throttle_netmasks) ... ok
test_external_nameserver (__main__.TestServicePowerDNS.test_external_nameserver) ... ok
test_listening_port (__main__.TestServicePowerDNS.test_listening_port) ... ok
test_no_rfc1918_forwarding (__main__.TestServicePowerDNS.test_no_rfc1918_forwarding) ... ok
test_serve_stale_extensions (__main__.TestServicePowerDNS.test_serve_stale_extensions) ... ok

----------------------------------------------------------------------
Ran 9 tests in 32.120s

OK

```
## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/service_dns_forwarding.xml.in [hunk 1]
```diff
@@ -694,6 +694,54 @@
...
+                  <valueHelp>
+                    <format>ipv4</format>
+                    <description>IPv4 address to match</description>
+                  </valueHelp>
+                  <valueHelp>
+                    <format>ipv4net</format>
+                    <description>IPv4 prefix to match</description>
+                  </valueHelp>
+                  <valueHelp>
+                    <format>!ipv4</format>
```
Comment (sever-sever) [inline interface-definitions/service_dns_forwarding.xml.in]: Did you test it?

```
set dont-throttle-netmasks 10.0.0.1
set dont-throttle-netmasks !10.0.0.2
set dont-throttle-netmasks !192.0.2.5
set dont-throttle-netmasks 10.0.0.2
```

Comment (fvlaicu) [reply interface-definitions/service_dns_forwarding.xml.in]:
```
vyos@gw-01# edit service dns forwarding
[edit service dns forwarding]
vyos@gw-01# set dont-throttle-netmasks 10.0.0.1
[edit service dns forwarding]
vyos@gw-01# set dont-throttle-netmasks !10.0.0.2
[edit service dns forwarding]
vyos@gw-01# set dont-throttle-netmasks !192.0.2.5
[edit service dns forwarding]
vyos@gw-01# set dont-throttle-netmasks 10.0.0.2
[edit service dns forwarding]
vyos@gw-01# compare
[forwarding]
+ dont-throttle-netmasks "10.0.0.1"
+ dont-throttle-netmasks "!10.0.0.2"
+ dont-throttle-netmasks "!192.0.2.5"
+ dont-throttle-netmasks "10.0.0.2"

[edit service dns forwarding]
vyos@gw-01# commit;save
[edit]
vyos@gw-01# show service dns forwarding do
domain                  dont-throttle-netmasks
[edit]
vyos@gw-01# show service dns forwarding dont-throttle-netmasks
 dont-throttle-netmasks 10.0.0.1
 dont-throttle-netmasks !10.0.0.2
 dont-throttle-netmasks !192.0.2.5
 dont-throttle-netmasks 10.0.0.2
[edit]
vyos@gw-01#

vyos@gw-01# cat /run/powerdns/recursor.conf
### Autogenerated by service_dns_forwarding.py ###

# XXX: pdns recursor doesn't like whitespace near entry separators,
# especially in the semicolon-separated lists of name servers.
# Please be careful if you edit the template.

# Non-configurable defaults
daemon=yes
threads=1
allow-from=xxxx
log-common-errors=yes
non-local-bind=yes
query-local-address=0.0.0.0,::
lua-config-file=recursor.conf.lua

# cache-size
max-cache-entries=1000000

# negative TTL for NXDOMAIN
max-negative-ttl=0

# timeout
network-timeout=1500

# ignore-hosts-file
export-etc-hosts=yes

# listen-address
local-address=xxxx

# listen-port
local-port=53

# dnssec
dnssec=off


# dont-throttle-netmasks
dont-throttle-netmasks=10.0.0.1,!10.0.0.2,!192.0.2.5,10.0.0.2
# serve rfc1918 records
serve-rfc1918=yes

# zones
auth-zones=
forward-zones-file=recursor.forward-zones.conf[edit]

```
However pdns doesn't respond anymore, trying on another instance to add a negated entry
```
root@gw-02:~# rec_control --socket-dir=/run/powerdns add-dont-throttle-netmasks '!10.0.0.3/32'
Problem parsing '!10.0.0.3/32': Unable to convert '!10.0.0.3' to a netmask, nothing added
```

Comment (fvlaicu) [reply interface-definitions/service_dns_forwarding.xml.in]: what is sad is that the official docs say that negation is supported: https://docs.powerdns.com/recursor/settings.html#dont-throttle-netmasks

but digging into git for the version that vyos is using: https://github.com/PowerDNS/pdns/blob/566ad636d52331435705fb3a9c876966e53a4d9d/pdns/recursordist/docs/settings.rst
it doesn't say anything about negation.
I will remove the negation part.

Comment (fvlaicu) [reply interface-definitions/service_dns_forwarding.xml.in]:
```
vyos@gw-01# sudo rec_control --socket-dir=/run/powerdns get-dont-throttle-netmasks
10.0.0.1/32, 10.0.0.2/32
```
after i removed the negated entries pdns responded :)

Cheers!

**Comment by sarthurdev** [comment]
Please update commit message in format shown here [CONTRIBUTING.md](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md#writing-good-commit-messages)

**Comment by c-po** [comment]
Please implement smoketests for every new feature

**Comment by sever-sever** [comment]
Read please [CONTRIBUTING.md](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md#writing-good-commit-messages) and recheck your commit messages

every change must be associated with a task number (prefixed with T) and a component. If there is no bug report/feature request for the changes you are going to make, you have to create a development portal task first. Once there is an entry in the development portal, you should reference its id in your commit message, as shown below:

`ddclient: T1030: auto create runtime directories`

**Review by sever-sever — APPROVED** [review]

File: interface-definitions/service_dns_forwarding.xml.in [hunk 2]
```diff
@@ -694,6 +707,34 @@
                   <valueless/>
                 </properties>
               </leafNode>
+              <leafNode name="dont-throttle-netmasks">
```
Comment (c-po) [inline interface-definitions/service_dns_forwarding.xml.in]: Sorry to be a bit picky here, but I do not like the `netmask` term.
Please rephrase this to: `exclude-throttle-address`

File: interface-definitions/service_dns_forwarding.xml.in [hunk 3]
```diff
@@ -670,6 +670,19 @@
                 </properties>
                 <defaultValue>3600</defaultValue>
               </leafNode>
+              <leafNode name="serve-stale-extensions">
```
Comment (c-po) [inline interface-definitions/service_dns_forwarding.xml.in]: Please use singular

File: interface-definitions/service_dns_forwarding.xml.in [hunk 4]
```diff
@@ -670,6 +670,19 @@
...
+                <properties>
+                  <help>Maximum number of times an expired record’s TTL is extended by 30s when serving stale</help>
```
Comment (c-po) [inline interface-definitions/service_dns_forwarding.xml.in]: Please avoid `’` in help string.

`Number of times an expired TTL record is extended by 30 seconds when serving stale`

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/service_dns_forwarding.xml.in (hunk 4 above)
Comment (fvlaicu) [reply interface-definitions/service_dns_forwarding.xml.in]: I went with 
```
Number of times the expired TTL of a record is extended by 30 seconds when serving stale
```

**Review by c-po — APPROVED** [review]

**Comment by c-po** [comment]
@MergifyIo backport sagitta

**Comment by c-po** [comment]
After the latest changes smoketests were not re-run. Builds broken.

Fixed via https://github.com/vyos/vyos-1x/pull/2761
